### DISCUSSION TITLE
How to register a Python func into a Rust struct and loop it?

### DISCUSSION THREAD

**Discussion by sun-rs** [PyO3/pyo3#2246] [Questions]
I have Some Rust code like below, it is a fixed rule to run. But I want to register some flexible Python func in it to achieve `Fixed System Struct + Flexible User Logic`. The Rust struct `App` has a `Run` func which takes in a struct that implemented `Context` trait. 

```rust
pub struct Wallet {
    pub cash: i64
}

impl Wallet {
    pub fn get_money(&mut self, num: i64) {
            self.cash += num;
    }
}


pub trait Context {
    /// execute py func here
    pub fn on_calculation(&mut self, wallet: &mut Wallet) {
          //CALL PYTHON METHOD HERE (need to send wallet)
          //pyclass.on_calculation(wallet) ?????
    }

    /// execute py func here
    pub fn finish(&mut self, wallet: &mut Wallet) {}
}

pub struct App {
    pyclass: pyclass
    wallet: Wallet,
}

impl App {

    pub fn run(&mut self, mut context: impl Context) {
        for _i in 0..1000000 {
            context(&mut self.wallet);
        }
        context.finish(&mut self.wallet);
    }
    pub fn reset_wallet(&mut self) {
        self.wallet.cash = 0;
    }
}

fn main() {
    struct TestStrategy {}
    impl Context for TestStrategy;
    let w = Wallet { cash:0 };
    let app = App { wallet:w };
    let stra = TestStrategy {};
    app.run(stra);
    app.reset_wallet();

}
```
While in Python , I have a class to execute the user code like below:
```python
class Context:
    def __init__():
        self.member = 999
    def on_calculation(self, wallet):
        if 1 > 0:
            wallet.get_money(10)
            
    def finish(self, wallet):
        print(wallet.cash)
```
How could I register the Python class method `def on_calculation` into Rust and interactive the result with Pyo3?  I couldn't find an example like that.

**Comment by birkenfeld** [comment]
You can store the function objects in the Context as `Py<PyAny>`. To call it, you need a `Python` token (usually passed as an argument).

If they are always methods you can also only store the object that has the methods and call them on it.

**Reply by sun-rs** [reply]
Thanks. Do you mean send whole Python class `Context` as a `Py<PyAny>` in Rust Struct, then write some wrapper func in Rust to call the method of `Py<PyAny>` which infact is a Python class? Just like [https://pyo3.rs/v0.16.2/trait_bounds.html](url). But how to send argument `wallet` to python method to use? I have update my example.

**Comment by mejrs** [comment]
Take a look at https://pyo3.rs/v0.16.2/class/call.html, it shows how to hold Python functions in Rust and how to call them.

**Reply by sun-rs** [reply]
Thanks for reply. I found an example may be fit for me. [[https://pyo3.rs/v0.16.2/trait_bounds.html](https://github.com/PyO3/pyo3/discussions/url)](url)

But I got confused of the example that if the python method (which has been send into rust struct as PyAny with its python class) need an argument which is a member of `struct UserModel` .Like my example before, the `wallet` need to be used in python method, now it is nested structure. How to solve it?
